File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/Syms/AttributeGroupReferencePersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.Syms
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Newtonsoft.Json.Linq;

    class AttributeGroupReferencePersistence
    {
        public static CdmAttributeGroupReference FromData(CdmCorpusContext ctx, JToken obj, string entityName = null)
        {
            if (obj == null)
            {
                return null;
            }
            bool simpleReference = true;
            dynamic attributeGroup;
            if (obj is JValue)
                attributeGroup = obj;
            else
            {
                simpleReference = false;
                if (obj["attributeGroupReference"] is JValue)
                    attributeGroup = (string)obj["attributeGroupReference"];
                else
                    attributeGroup = AttributeGroupPersistence.FromData(ctx, obj["attributeGroupReference"], entityName);
            }

            return ctx.Corpus.MakeRef<CdmAttributeGroupReference>(CdmObjectType.AttributeGroupRef, attributeGroup, simpleReference);
        }
        public static dynamic ToData(CdmAttributeGroupReference instance, ResolveOptions resOpt, CopyOptions options)
        {
            return CdmObjectRefPersistence.ToData(instance, resOpt, options);
        }

    }
}

File: objectModel/Python/cdm/persistence/syms/attribute_group_reference_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Optional, Union, TYPE_CHECKING

from cdm.enums import CdmObjectType

from .attribute_group_persistence import AttributeGroupPersistence
from .cdm_object_ref_persistence import CdmObjectRefPersistence

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusContext, CdmAttributeGroupReference

    from .types import AttributeGroupReference


class AttributeGroupReferencePersistence(CdmObjectRefPersistence):
    @staticmethod
    def from_data(ctx: 'CdmCorpusContext', data: Union[str, 'AttributeGroupReference'], entity_name: Optional[str] = None) -> Optional['CdmAttributeGroupReference']:
        if not data:
            return None

        simple_reference = True
        attribute_group = None

        if isinstance(data, str):
            attribute_group = data
        else:
            simple_reference = False
            if isinstance(data.attributeGroupReference, str):
                attribute_group = data.attributeGroupReference
            else:
                attribute_group = AttributeGroupPersistence.from_data(ctx, data.attributeGroupReference, entity_name)

        return ctx.corpus.make_ref(CdmObjectType.ATTRIBUTE_GROUP_REF, attribute_group, simple_reference)
